tlatools/org.lamport.tlatools/src/tlc2/REPL.java

```diff
@@ -0,0 +1,166 @@
+            lines.add(replValueVarName + " == " + evalExpr);
+            lines.add("====");
+
+            // Write out the spec file.
+            BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile.getAbsolutePath(), false));
+            for (String line : lines) {
+                writer.append(line);
+                writer.newLine();
+            }
+            writer.close();
+
+            // Avoid sending log messages to stdout and reset the messages recording.
+            ToolIO.setMode(ToolIO.TOOL);
+            ToolIO.reset();
+
+            try {
+                // We placed the REPL spec files into a temporary directory, so, we add this temp directory
+                // path to the filename resolver used by the Tool.
+                SimpleFilenameToStream resolver = new SimpleFilenameToStream(replTempDir.toAbsolutePath().toString());
+                Tool tool = new FastTool(REPL_SPEC_NAME, REPL_SPEC_NAME, resolver);
+                ModuleNode module = tool.getSpecProcessor().getRootModule();
+                OpDefNode valueNode = module.getOpDef(replValueVarName);
+                Value exprVal = (Value) tool.eval(valueNode.getBody(), Context.Empty, TLCState.Empty);
+                return exprVal.toString();
+            } catch (Assert.TLCRuntimeException | EvalException exc) {
+                // TODO: Improve error messages with more specific detail.
+                System.out.println("Error evaluating expression: '" + evalExpr + "'");
+            }
+
+        } catch (IOException pe) {
+            pe.printStackTrace();
+        }
+        return "";
+    }
+
+    /**
+     * Runs the main REPL loop continuously until there is a fatal error or a user interrupt.
+     */
+    public void runREPL() throws IOException {
+        // For TLA+ we don't want to treat backslashes as escape chars e.g. for LaTeX like operators.
+        DefaultParser parser = new DefaultParser();
+        parser.setEscapeChars(null);
+        Terminal terminal = TerminalBuilder.builder().build();
+        LineReader reader = LineReaderBuilder.builder().parser(parser).terminal(terminal).build();
+
+        // Run the loop.
+        String prompt = "(tla+) ";
+        String expr;
+        while (true) {
+            try {
+                expr = reader.readLine(prompt);
+                String res = processInput(expr);
+                if (res.equals("")) {
+                    continue;
+                }
+                System.out.println(res);
+            } catch (UserInterruptException e) {
+                return;
+            } catch (EndOfFileException e) {
+                e.printStackTrace();
+                return;
+            }
+        }
+    }
+
+    public static void main(String[] args) {
+        try {
+            Path tempDir = Files.createTempDirectory(TEMP_DIR_PREFIX);
+            REPL repl = new REPL(tempDir);
+            // TODO: Allow external spec file to be loaded into REPL context.
```

Comment:
Is the REPL supposed to replace the Toolbox's Evaluate Constant Expression functionality for command-line users (this is when users would want to load a spec)?  If yes, how does a user define constants?

Reply:
That's a good point. For now, I was viewing this initial REPL implementation as mostly a built in TLC replacement for the functionality of https://github.com/will62794/tlaplus_repl, since that seems have to been useful for others when first exploring the language. That only provides the ability to evaluate expressions without any loaded TLA+ module. To support loading modules, I guess we would need to require the user give a config, same as they would when checking their spec `tlc2.TLC`. I forgot that in my comment above.

Ideally this REPL implementation can serve as a foundation for both CLI use cases and integration into other tools, including the Toolbox and/or VS Code. My thought was that for users who prefer to play around in a REPL from the CLI, they can use this tool, but that we could also take advantage of the low latency evaluation capability for improving things like the "Evaluate Constant Expression" feature of the Toolbox, and the expression evaluation feature of the VS Code extension. That might require having some sort of simple interprocess communication mechanism (plaintext?), if we wanted to have the Toolbox, for example, keep a REPL process running in the background that it could use for quick evaluation of an expression in context of a spec/model.

What are your thoughts on this?